symbolic-base/src/ZkFold/Base/Data/Eq.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE DerivingVia   #-}
+{-# LANGUAGE TypeOperators #-}
+
+module ZkFold.Base.Data.Eq
+  ( Eq (..)
+  , elem
+  , GEq (..)
+  ) where
+
+import           Data.Foldable                    hiding (all, and, any, elem, or)
+import           Data.Functor.Identity
+import qualified Data.Functor.Rep                 as R
+import qualified GHC.Generics                     as G
+import           Prelude                          (type (~))
+import qualified Prelude                          as H
+
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Data.Boolean
+import           ZkFold.Base.Data.Conditional
+import           ZkFold.Base.Data.Vector
+
+class Conditional a => Eq a where
+
+  (==) :: a -> a -> BooleanOf a
+  default (==) ::
+    ( G.Generic a
+    , GEq (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => a -> a -> BooleanOf a
+  x == y = geq (G.from x) (G.from y)
+
+  (/=) :: a -> a -> BooleanOf a
```

Comment:
Well, if `(/=)` is defined like this, then the definition from `GEq` is not used at all
Sorry for the previous comment, let's keep `(/=)` as you defined it for now